src/data/real/dedekind_cut.lean

```diff
@@ -0,0 +1,403 @@
+/-
+Copyright (c) 2022 Yakov Pechersky. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Yakov Pechersky
+-/
+import algebra.order.archimedean
+import data.set.pointwise
+import order.upper_lower
+
+/-!
+# Dedeking cuts
+
+A basic theory of Dedekind cuts, used in a construction of the reals.
+
+## Important definitions
+
+- `dedekind_cut Q` is an `add_comm_group` when `Q` is `linear_ordered_field` and `archimedean`
+- TODO: finish constructing the full field
+
+## Tags
+
+dedekind, real, cut
+
+## References
+
+[Rudin, *Principles of Mathematical Analysis* (Chapter 1, Appendix)][rudin1976]
+
+-/
+
+section
+
+variables (Q : Type*) [has_lt Q]
+
+/-- *Dedekind cut*, defined for an arbitrary type with a less-than relation -/
+@[ext] structure dedekind_cut :=
+(carrier : set Q)
+(nonempty' : carrier.nonempty) -- Step 1, (I), first half
+(ne_top' : carrier ≠ ⊤) -- Step 1, (I), second half
```

Comment:
I'm 90% sure that `set.univ` is preferred to using `⊤`, someone should confirm or deny this.